scraper/src/libretexts2zim/css.py

```diff
@@ -0,0 +1,140 @@
+        css_assets_root_path: Path = Path("/content/css_assets"),
+    ) -> None:
+        self.css_target_path = css_target_path
+        self.css_assets_root_path = css_assets_root_path
+        self.css_assets: dict[OriginalUrl, FullZimPath] = {}
+        self.used_paths: list[RelativeCssPath] = []
+
+    def process(self, css_original_url: str, css_content: bytes) -> str:
+        rules, _ = parse_stylesheet_bytes(  # pyright: ignore[reportUnknownVariableType]
+            css_content
+        )
+        self._process_list(
+            css_original_url,
+            rules,  # pyright: ignore[reportUnknownArgumentType]
+        )
+        return serialize(rules)
+
+    def _process_url(
+        self, css_original_url: str, css_url: str
+    ) -> RelativeCssPath | None:
+        original_url = urljoin(css_original_url, css_url)
+        original_url_parsed = urlparse(original_url)
+        if original_url_parsed.scheme.lower() not in ["http", "https"]:
+            return None
+        if original_url in self.css_assets:
+            return self.css_assets[original_url].relative_to(self.css_target_path)
+        original_path = Path(urlparse(original_url).path)
+        target_parent = Path(
+            *[
+                parent.name
+                for parent in reversed(original_path.parents)
+                if parent.name and parent.name != ".."
+            ]
+        )
+
+        index = 0
+        while True:
+            relative_path = (
+                target_parent
+                / f"{original_path.stem}{'_' + str(index) if index else ''}"
+                f"{original_path.suffix}"
+            )
+            if relative_path not in self.used_paths:
+                break
+            index += 1
+
+        self.used_paths.append(relative_path)
+        target_path = self.css_assets_root_path / relative_path
+        self.css_assets[original_url] = target_path
+        return target_path.relative_to(self.css_target_path)
+
+    def _process_node(self, css_original_url: str, node: ast.Node):
+        if isinstance(
+            node,
+            ast.QualifiedRule
+            | ast.SquareBracketsBlock
+            | ast.ParenthesesBlock
+            | ast.CurlyBracketsBlock,
+        ):
+            self._process_list(
+                css_original_url,
+                node.content,  # pyright: ignore[reportUnknownArgumentType, reportUnknownMemberType]
+            )
+        elif isinstance(node, ast.FunctionBlock):
+            if node.lower_name == "url":  # pyright: ignore[reportUnknownMemberType]
+                url_node: ast.Node = node.arguments[0]  # pyright: ignore
+                relative_css_path = self._process_url(
+                    css_original_url,
+                    url_node.value,  # pyright: ignore
+                )
+                if not relative_css_path:
+                    return
+                url_node.value = str(relative_css_path)  # pyright: ignore
```

Comment:
Is there a reason for those wide ignores?

Reply:
The reason is that there is 2 or 3 rules to ignore, I don't recall exactly, but when you specify it then it doesn't fit on one line, so black want to move things to match the 88 chars limit, and then it becomes a mess because hint is not a proper line anymore, ... Not ideal obviously.

Reply:
Yeah I hate those. The problem is that those will never be fixed because the next developer will have no clue what was being ignored and this will lead to more stuff ignored, eventually weakening what we want to rely on.

I propose we keep them and disable black for those specific lines.

Here's what it would look like on that function:

```py
    def _process_node(self, css_original_url: str, node: ast.Node):
        """Process one single CSS node"""
        if isinstance(
            node,
            ast.QualifiedRule
            | ast.SquareBracketsBlock
            | ast.ParenthesesBlock
            | ast.CurlyBracketsBlock,
        ):
            self._process_list(
                css_original_url,
                node.content,  # pyright: ignore[reportUnknownArgumentType, reportUnknownMemberType]
            )
        elif isinstance(node, ast.FunctionBlock):
            if node.lower_name == "url":  # pyright: ignore[reportUnknownMemberType]
                url_node: ast.Node = node.arguments[0]  # pyright: ignore
                relative_css_path = self._process_url(
                    css_original_url,
                    url_node.value,  # pyright: ignore [reportAttributeAccessIssue]
                )
                if not relative_css_path:
                    return
                url_node.value = str(relative_css_path)  # pyright: ignore [reportAttributeAccessIssue] ; fmt: skip
                url_node.representation = f'"{serialize_url(str(relative_css_path))}"'  # pyright: ignore [reportAttributeAccessIssue] ; fmt: skip

            else:
                self._process_list(
                    css_original_url,
                    node.arguments,
                )
        elif isinstance(node, ast.AtRule):
            self._process_list(
                css_original_url,
                node.prelude,
            )
            self._process_list(
                css_original_url,
                node.content,
            )
        elif isinstance(node, ast.Declaration):
            self._process_list(
                css_original_url,
                node.value,
            )
        elif isinstance(node, ast.URLToken):
            relative_css_path = self._process_url(
                css_original_url,
                node.value,
            )
            if not relative_css_path:
                return
            node.value = str(relative_css_path)
            node.representation = f"url({serialize_url(str(relative_css_path))})"

```

Interestingly, I simply removed three of those ignores 🙃